apps/dav/lib/Upload/ChunkingV2Plugin.php

```diff
@@ -0,0 +1,392 @@
+			// the storage/path used in getStorage()
+			if ($actualFile instanceof File && $this->uploadFolder->getStorage()->getId() === $actualFile->getNode()->getStorage()->getId()) {
+				return $actualFile;
+			}
+		} catch (NotFound $e) {
+			// If there is no target file we upload to the upload folder first
+		}
+
+		// Use file in the upload directory that will be copied or moved afterwards
+		if ($createIfNotExists) {
+			$this->uploadFolder->createFile(self::TEMP_TARGET);
+		}
+
+		/** @var UploadFile $uploadFile */
+		$uploadFile = $this->uploadFolder->getChild(self::TEMP_TARGET);
+		return $uploadFile->getFile();
+	}
+
+	public function afterMkcol(RequestInterface $request, ResponseInterface $response): bool {
+		try {
+			$this->prepareUpload($request->getPath());
+			$this->checkPrerequisites(false);
+		} catch (BadRequest|StorageInvalidException|NotFound $e) {
+			return true;
+		}
+
+		$this->uploadPath = $this->server->calculateUri($this->server->httpRequest->getHeader(self::DESTINATION_HEADER));
+		$targetFile = $this->getUploadFile($this->uploadPath, true);
+		[$storage, $storagePath] = $this->getUploadStorage($this->uploadPath);
+
+		$this->uploadId = $storage->startChunkedWrite($storagePath);
+
+		$this->cache->set($this->uploadFolder->getName(), [
+			self::UPLOAD_ID => $this->uploadId,
+			self::UPLOAD_TARGET_PATH => $this->uploadPath,
+			self::UPLOAD_TARGET_ID => $targetFile->getId(),
+		], 86400);
+
+		$response->setStatus(201);
+		return true;
+	}
+
+	public function beforePut(RequestInterface $request, ResponseInterface $response): bool {
+		try {
+			$this->prepareUpload(dirname($request->getPath()));
+			$this->checkPrerequisites();
+		} catch (StorageInvalidException|BadRequest|NotFound $e) {
+			return true;
+		}
+
+		[$storage, $storagePath] = $this->getUploadStorage($this->uploadPath);
+
+		$chunkName = basename($request->getPath());
+		$partId = is_numeric($chunkName) ? (int)$chunkName : -1;
+		if (!($partId >= 1 && $partId <= 10000)) {
+			throw new BadRequest('Invalid chunk name, must be numeric between 1 and 10000');
+		}
+
+		$uploadFile = $this->getUploadFile($this->uploadPath);
+		$tempTargetFile = null;
+
+		$additionalSize = (int)$request->getHeader('Content-Length');
+		if ($this->uploadFolder->childExists(self::TEMP_TARGET) && $this->uploadPath) {
+			/** @var UploadFile $tempTargetFile */
+			$tempTargetFile = $this->uploadFolder->getChild(self::TEMP_TARGET);
+			[$destinationDir, $destinationName] = Uri\split($this->uploadPath);
+			/** @var Directory $destinationParent */
+			$destinationParent = $this->server->tree->getNodeForPath($destinationDir);
+			$free = $storage->free_space($destinationParent->getInternalPath());
+			$newSize = $tempTargetFile->getSize() + $additionalSize;
+			if ($free >= 0 && ($tempTargetFile->getSize() > $free || $newSize > $free)) {
+				throw new InsufficientStorage("Insufficient space in $this->uploadPath");
```

Comment:
```suggestion
				throw new InsufficientStorage("Insufficient space in {$this->uploadPath}");
```
(personal readability preference, anything more complex than $var should go into {})